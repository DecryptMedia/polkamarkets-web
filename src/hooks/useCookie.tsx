// https://medium.com/swlh/react-hooks-usecookie-hook-26ac06ff36b0

import { useCallback } from 'react';

import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';

dayjs.extend(utc);

const getItem = key =>
  document.cookie.split('; ').reduce((total, currentCookie) => {
    const [storedKey, storedValue] = currentCookie.split('=');

    return key === storedKey ? decodeURIComponent(storedValue) : total;
  }, '');

const setItem = (key, value) => {
  const now = dayjs();
  const expiresAt = now.add(1, 'hour');

  document.cookie = `${key}=${value}; expires=${expiresAt
    .utc()
    .toString()}; path=/`;
};

export default function useCookie(
  key,
  defaultValue
): [string, (_value: unknown) => void] {
  const cookie = getItem(key);
  const setCookie = useCallback(value => setItem(key, value), [key]);

  setCookie(defaultValue);

  return [cookie, setCookie];
}
