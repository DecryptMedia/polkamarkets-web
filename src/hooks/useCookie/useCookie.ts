import { useCallback, useMemo } from 'react';

import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';

dayjs.extend(utc);

/**
 * Hook with cookie getter and setter, that uses the Document interface object.
 * @param {string} key The cookie name.
 * @param {string} defaultValue The default value that will be returned if no cookie was found with the provided cookie's name.
 * @returns {[string | null, (newValue: string) => void]} An array with the first value being the returned cookie or null, and the second as a callback for a setter of a new cookie.
 */
export default function useCookie<V>(key: string, defaultValue?: V) {
  const value = useMemo(
    () =>
      document.cookie?.split('; ').reduce((total, currentCookie) => {
        const [storedKey, storedValue] = currentCookie.split('=');

        return key === storedKey ? decodeURIComponent(storedValue) : total;
      }, ''),
    [key]
  );
  const setCookie = useCallback(
    (newValue: V) => {
      const expiresAt = dayjs().add(1, 'hour').utc().toString();

      document.cookie = `${key}=${newValue}; expires=${expiresAt}; path=/`;
    },
    [key]
  );

  return [value || defaultValue || null, setCookie] as const;
}
