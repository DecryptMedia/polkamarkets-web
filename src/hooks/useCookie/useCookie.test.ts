import { act, renderHook } from '@testing-library/react-hooks';
import _dayjs from 'dayjs';
import mockDate from 'mockdate';

import useCookie from './useCookie';

enum Cookie {
  Key = 'KEY',
  Value = 'VALUE'
}

function renderCookie<V>(key: string, defaultValue?: V) {
  return renderHook(
    initialProps => useCookie(initialProps.key, initialProps.defaultValue),
    { initialProps: { key, defaultValue } }
  );
}

beforeEach(() => {
  Object.defineProperty(document, 'cookie', {
    writable: true,
    value: null
  });
  mockDate.set('1983-01-01');
});
afterEach(() => {
  mockDate.reset();
});
describe('useCookie', () => {
  it('returns null for non-stored cookie', () => {
    const { result } = renderCookie(Cookie.Key);
    const [cookie] = result.current;

    expect(cookie).toBeNull();
  });
  it('returns a stored cookie', () => {
    document.cookie = `${Cookie.Key}=${Cookie.Value}; `;

    const { result } = renderCookie(Cookie.Key);
    const [cookie] = result.current;

    expect(cookie).toBe(Cookie.Value);
  });
  it('returns a default cookie', () => {
    const { result } = renderCookie(Cookie.Key, Cookie.Value);
    const [cookie] = result.current;

    expect(cookie).toBe(Cookie.Value);
  });
  it('sets a cookie', () => {
    const { result } = renderCookie(Cookie.Key);
    const [cookie, setCookie] = result.current;

    expect(cookie).toBeNull();
    expect(typeof setCookie).toBe('function');
    act(() => setCookie(Cookie.Value));
    expect(document.cookie).toMatchSnapshot();
    expect(document.cookie).toContain(`${Cookie.Key}=${Cookie.Value};`);
  });
});
