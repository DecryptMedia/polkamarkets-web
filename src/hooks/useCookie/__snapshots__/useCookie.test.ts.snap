// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCookie sets a cookie 1`] = `"KEY=VALUE; expires=Sat, 01 Jan 1983 01:00:00 GMT; path=/"`;
